### PR TITLE
Upgrade Rails to v7.1.3.2

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

:gem: **We're upgrading Rails from v7 to v7.1** :steam_locomotive:

Release Notes: https://guides.rubyonrails.org/7_1_release_notes.html
Upgrade Guide: https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-7-0-to-rails-7-1
Incremental upgrade options: [https://github.com/rails/rails/blob/6f0d1ad14b92b9f5906e44740fce8b4f1c7075dc/railt[…]/templates/config/initializers/new_framework_defaults_7_1.rb.tt](https://github.com/rails/rails/blob/6f0d1ad14b92b9f5906e44740fce8b4f1c7075dc/railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_1.rb.tt)

## Changes proposed in this pull request

This is a minor version upgrade of the Rails framework.

1. **Zeitwerk now autoloads the lib directory**

	Add `config.autoload_lib(ignore: %w(assets tasks generators rubocop))`
	`AWSIpRanges` needs changes because Zeitwerk doesn't like how it's inflected.

2. **`config.cache_classes = true` is now `config.enable_reloading = false`**

3. **Running `bin/rails app:update` generates double quotes, make that default**

4. **`ActionDispatch::HostAuthorization` middleware is not guaranteed to be present**

	We don't populate the `config.hosts` property in production so the middleware is not used.
	We need to choose another middleware to insert middlewares relative to.

5. **`form_for` and `form_with` refactoring in Rails caused a bug in RejectionsWizard**

	Remove the implementation of `#to_model` in the `RejectionsWizard`. Implementing `#to_model` to return the underlying object puts too many dependencies on Rails internals.

	**Rails 7..7.1 fixes for `#to_model`**
	https://github.com/rails/rails/pull/43421
	https://github.com/rails/rails/issues/44326
	https://github.com/rails/rails/pull/44468

6. **Rails default sanitizer updated to Rails::HTML5::Sanitizer**

	This feature is in it's own gem https://github.com/rails/rails-html-sanitizer
	Option updated in config https://github.com/rails/rails/blob/6f0d1ad14b92b9f5906e44740fce8b4f1c7075dc/railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_1.rb.tt#L254
	
	For some reason we had to move `ActionView::Base.sanitized_allowed_tags` to
	`config.action_view.sanitized_allowed_tags` otherwise the sanitized was not stripping all the tags.

#### Options to enable?

**Annotate rendered view with file names**
```ruby
  # Annotate rendered view with file names.
  config.action_view.annotate_rendered_view_with_filenames = true
```
Screenshot from 2024-03-12 12-17-17

## Link to Trello card

[Trello Ticket](https://trello.com/c/WYjfFRQx/1393-upgrading-rails-to-71)

### CODE REVIEW

File: config/initializers/permissions_policy.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Be sure to restart your server when you modify this file.
```
Comment (elceebee) [inline config/initializers/permissions_policy.rb:1]: Does this file need to be included if it's all commented out?

**Review by elceebee — APPROVED** [review]
Great work, these upgrades can be so fiddly and a bit mysterious at times. 

Running it locally, there is just one thing really. When I ran `bin/setup`, it made this change in the schema. I think you will probably want to do the same and include change with your PR:

image

Otherwise, everything looks great. Thanks for doing this!

File: config/initializers/permissions_policy.rb (hunk 1 above)
Comment (inulty-dfe) [reply config/initializers/permissions_policy.rb:1]: It's autogenerated and I prefer to keep the autogenerated files even if they're commented out.

It's brought our attention to one thing at least that we could consider valuable. =>

https://securityheaders.com/?q=https%3A%2F%2Fwww.apply-for-teacher-training.service.gov.uk%2F&followRedirects=on

**Comment by inulty-dfe** [comment]
[Upgrading Rails to 7.1](https://trello.com/c/WYjfFRQx/1393-upgrading-rails-to-71)

File: config/application.rb [hunk 2]
```diff
@@ -1,65 +1,73 @@
...
 
 require_relative "../lib/modules/aws_ip_ranges"
 
-require 'grover'
+require "grover"
 
 module ApplyForPostgraduateTeacherTraining
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 7.0
```
Comment (inulty-dfe) [inline config/application.rb:37]: This should be updated to 7.1 

or 

we use the incremental upgrade file.

**Review by dcyoung-dev — APPROVED** [review]

**Review by tomas-stefano — APPROVED** [review]
Code looks good. Tomorrow I will play around with the review app ✌️

**Comment by tomas-stefano** [comment]
Tested the review app, overall I couldn't identify anything wrong. All good :)

File: Gemfile [hunk 3]
```diff
@@ -58,7 +58,8 @@ gem 'oj'
 # This gem adds support for sequences in the schema.rb
 gem 'ar-sequence'
 
-gem 'active_hash'
+# Source gem from rubygems after ActiveHash version > 3.2.1
+gem 'active_hash', github: 'active-hash/active_hash', ref: '195b1e9'
```
[the remark this answers is not in the document]
Comment (inulty-dfe) [reply Gemfile:62]: Good point. 

This approach will not pull in the commits that have been made up to `ref: '195b1e9'` which have not been released yet.

**Pull request merged**